File: python/conflictgraph/scheduler.py
from __future__ import annotations

import heapq
import random
import time
import uuid
from dataclasses import dataclass
from typing import Iterable, Mapping, Optional, Sequence

from .types import (
    PairPrediction,
    RiskPolicy,
    Schedule,
    ScheduledTest,
    TestIdentity,
    TestStats,
    stable_pair,
)


@dataclass(frozen=True)
class PolicyParameters:
    risk_weight: float
    hard_threshold: float
    utilization_bias: float
    lookahead: int


POLICIES: dict[RiskPolicy, PolicyParameters] = {
    RiskPolicy.AGGRESSIVE: PolicyParameters(0.55, 0.985, 1.5, 8),
    RiskPolicy.BALANCED: PolicyParameters(2.0, 0.90, 1.0, 16),
    RiskPolicy.SAFE: PolicyParameters(7.5, 0.68, 0.5, 32),
}


class RiskIndex:
    def __init__(self, predictions: Iterable[PairPrediction]) -> None:
        self._risks: dict[tuple[str, str], PairPrediction] = {
            item.key: item for item in predictions
        }

    def probability(self, left: str, right: str) -> float:
        if left == right:
            return 1.0
        prediction = self._risks.get(stable_pair(left, right))
        return prediction.probability if prediction else 0.0

    def prediction(self, left: str, right: str) -> Optional[PairPrediction]:
        return self._risks.get(stable_pair(left, right))


class ConflictScheduler:
    """Duration-aware list scheduler with conflict costs and deterministic refinement."""

    def __init__(
        self,
        workers: int,
        policy: RiskPolicy = RiskPolicy.BALANCED,
        seed: int = 42,
        risk_weight: Optional[float] = None,
        hard_threshold: Optional[float] = None,
        refinement_rounds: int = 4,
    ) -> None:
        if workers <= 0:
            raise ValueError("workers must be positive")
        self.workers = workers
        self.policy = policy
        self.seed = seed
        base = POLICIES[policy]
        self.params = PolicyParameters(
            risk_weight if risk_weight is not None else base.risk_weight,
            hard_threshold if hard_threshold is not None else base.hard_threshold,
            base.utilization_bias,
            base.lookahead,
        )
        self.refinement_rounds = refinement_rounds

    def static_schedule(
        self,
        tests: Sequence[TestIdentity],
        stats: Mapping[str, TestStats],
        predictions: Iterable[PairPrediction],
        run_id: Optional[str] = None,
    ) -> Schedule:
        started = time.perf_counter()
        risk = RiskIndex(predictions)
        randomizer = random.Random(self.seed)
        # Seeded tie-break ensures determinism without always favoring lexicographic IDs.
        tie = {test.id: randomizer.random() for test in tests}
        durations = {
            test.id: max(0.001, stats.get(test.id, TestStats()).duration_ema) for test in tests
        }
        by_id = {test.id: test for test in tests}
        remaining = sorted(
            by_id,
            key=lambda test_id: (-durations[test_id], tie[test_id], test_id),
        )
        lanes: list[list[ScheduledTest]] = [[] for _ in range(self.workers)]
        lane_ends = [0.0] * self.workers

        while remaining:
            worker = min(range(self.workers), key=lambda index: (lane_ends[index], index))
            start = lane_ends[worker]
            running = self._overlapping_at(lanes, start, worker)
            candidates = remaining[: self.params.lookahead]
            chosen = min(
                candidates,
                key=lambda test_id: self._placement_cost(
                    test_id, start, durations[test_id], lanes, risk, worker
                ),
            )
            probability, reasons = self._placement_risk(
                chosen, start, start + durations[chosen], lanes, risk, worker
            )

            # A hard-risk candidate may leave the worker briefly idle until the earliest
            # conflicting test finishes, provided another safe candidate is unavailable.
            if probability >= self.params.hard_threshold and running:
                safe = [
                    candidate
                    for candidate in candidates
                    if self._placement_risk(
                        candidate, start, start + durations[candidate], lanes, risk, worker
                    )[0]
                    < self.params.hard_threshold
                ]
                if safe:
                    chosen = min(
                        safe,
                        key=lambda test_id: self._placement_cost(
                            test_id, start, durations[test_id], lanes, risk, worker
                        ),
                    )
                    probability, reasons = self._placement_risk(
                        chosen, start, start + durations[chosen], lanes, risk, worker
                    )
                else:
                    next_end = min(
                        item.estimated_end
                        for lane_index, lane in enumerate(lanes)
                        if lane_index != worker
                        for item in lane
                        if item.estimated_start <= start < item.estimated_end
                    )
                    start = max(start, next_end)
                    probability, reasons = self._placement_risk(
                        chosen, start, start + durations[chosen], lanes, risk, worker
                    )
            scheduled = ScheduledTest(
                chosen,
                by_id[chosen].node_id,
                worker,
                start,
                start + durations[chosen],
                durations[chosen],
                probability,
                reasons,
            )
            lanes[worker].append(scheduled)
            lane_ends[worker] = scheduled.estimated_end
            remaining.remove(chosen)

        # Refinement is quadratic per attempted swap. Bounded list scheduling is
        # used directly for large suites so the production path stays responsive.
        if len(tests) <= 64:
            self._refine(lanes, risk)
        flattened = sorted(
            (item for lane in lanes for item in lane),
            key=lambda item: (item.estimated_start, item.worker, item.test_id),
        )
        expected_risk = self._schedule_risk(flattened, risk)
        elapsed_ms = (time.perf_counter() - started) * 1000
        return Schedule(
            id=str(uuid.uuid4()),
            run_id=run_id or str(uuid.uuid4()),
            workers=self.workers,
            policy=self.policy,
            tests=flattened,
            scheduler_latency_ms=elapsed_ms,
            expected_makespan=max((item.estimated_end for item in flattened), default=0.0),
            expected_risk=expected_risk,
            seed=self.seed,
        )

    def _placement_cost(
        self,
        test_id: str,
        start: float,
        duration: float,
        lanes: list[list[ScheduledTest]],
        risk: RiskIndex,
        worker: int,
    ) -> float:
        probability, _ = self._placement_risk(test_id, start, start + duration, lanes, risk, worker)
        # LPT pressure starts long tail work early; conflict exposure can still make a
        # shorter safe test preferable when a long candidate would overlap real risk.
        return (
            start
            - self.params.utilization_bias * duration
            + self.params.risk_weight * probability * duration
        )

    @staticmethod
    def _overlapping_at(
        lanes: list[list[ScheduledTest]], instant: float, excluded_worker: int
    ) -> list[ScheduledTest]:
        return [
            item
            for index, lane in enumerate(lanes)
            if index != excluded_worker
            for item in lane
            if item.estimated_start <= instant < item.estimated_end
        ]

    def _placement_risk(
        self,
        test_id: str,
        start: float,
        end: float,
        lanes: list[list[ScheduledTest]],
        risk: RiskIndex,
        worker: int,
    ) -> tuple[float, list[str]]:
        safe = 1.0
        reasons: list[str] = []
        for index, lane in enumerate(lanes):
            if index == worker:
                continue
            # Lane end times are monotonic. Skip completed work with binary search,
            # then inspect only the tests that actually overlap this placement.
            low, high = 0, len(lane)
            while low < high:
                middle = (low + high) // 2
                if lane[middle].estimated_end <= start:
                    low = middle + 1
                else:
                    high = middle
            for other in lane[low:]:
                if other.estimated_start >= end:
                    break
                overlap = max(
                    0.0, min(end, other.estimated_end) - max(start, other.estimated_start)
                )
                if overlap <= 0:
                    continue
                probability = risk.probability(test_id, other.test_id)
                exposure = min(
                    1.0, overlap / max(0.001, min(end - start, other.estimated_duration))
                )
                safe *= 1.0 - probability * exposure
                if probability >= 0.25:
                    prediction = risk.prediction(test_id, other.test_id)
                    explanation = (
                        prediction.explanation if prediction else "predicted shared-resource risk"
                    )
                    reasons.append(f"{other.node_id}: {probability:.1%} ({explanation})")
        return 1.0 - safe, reasons

    def _refine(self, lanes: list[list[ScheduledTest]], risk: RiskIndex) -> None:
        """Perform deterministic adjacent swaps that strictly improve makespan+risk."""
        if self.refinement_rounds <= 0:
            return
        for _ in range(self.refinement_rounds):
            changed = False
            baseline = self._objective([item for lane in lanes for item in lane], risk)
            for lane_index, lane in enumerate(lanes):
                for index in range(len(lane) - 1):
                    lane[index], lane[index + 1] = lane[index + 1], lane[index]
                    self._reflow_lane(lane, lane_index)
                    candidate = self._objective([item for group in lanes for item in group], risk)
                    if candidate + 1e-9 < baseline:
                        baseline, changed = candidate, True
                    else:
                        lane[index], lane[index + 1] = lane[index + 1], lane[index]
                        self._reflow_lane(lane, lane_index)
            if not changed:
                break

    @staticmethod
    def _reflow_lane(lane: list[ScheduledTest], worker: int) -> None:
        cursor = 0.0
        for item in lane:
            item.worker = worker
            item.estimated_start = cursor
            item.estimated_end = cursor + item.estimated_duration
            cursor = item.estimated_end

    def _objective(self, tests: list[ScheduledTest], risk: RiskIndex) -> float:
        makespan = max((item.estimated_end for item in tests), default=0.0)
        return makespan + self.params.risk_weight * self._schedule_risk(tests, risk)

    @staticmethod
    def _schedule_risk(tests: Sequence[ScheduledTest], risk: RiskIndex) -> float:
        total = 0.0
        for index, left in enumerate(tests):
            for right in tests[index + 1 :]:
                if left.worker == right.worker:
                    continue
                overlap = max(
                    0.0,
                    min(left.estimated_end, right.estimated_end)
                    - max(left.estimated_start, right.estimated_start),
                )
                if overlap:
                    total += risk.probability(left.test_id, right.test_id) * overlap
        return total


class RandomScheduler:
    def __init__(self, workers: int, seed: int = 42) -> None:
        self.workers, self.seed = workers, seed

    def schedule(
        self,
        tests: Sequence[TestIdentity],
        stats: Mapping[str, TestStats],
        run_id: Optional[str] = None,
    ) -> Schedule:
        started = time.perf_counter()
        order = list(tests)
        random.Random(self.seed).shuffle(order)
        heap = [(0.0, worker) for worker in range(self.workers)]
        lanes: list[ScheduledTest] = []
        for test in order:
            start, worker = heapq.heappop(heap)
            duration = max(0.001, stats.get(test.id, TestStats()).duration_ema)
            lanes.append(
                ScheduledTest(test.id, test.node_id, worker, start, start + duration, duration)
            )
            heapq.heappush(heap, (start + duration, worker))
        elapsed = (time.perf_counter() - started) * 1000
        return Schedule(
            str(uuid.uuid4()),
            run_id or str(uuid.uuid4()),
            self.workers,
            RiskPolicy.AGGRESSIVE,
            lanes,
            elapsed,
            max((x.estimated_end for x in lanes), default=0),
            0,
            self.seed,
        )


class SerialScheduler:
    def schedule(
        self,
        tests: Sequence[TestIdentity],
        stats: Mapping[str, TestStats],
        run_id: Optional[str] = None,
    ) -> Schedule:
        started = time.perf_counter()
        cursor = 0.0
        scheduled: list[ScheduledTest] = []
        for test in tests:
            duration = max(0.001, stats.get(test.id, TestStats()).duration_ema)
            scheduled.append(
                ScheduledTest(test.id, test.node_id, 0, cursor, cursor + duration, duration)
            )
            cursor += duration
        return Schedule(
            str(uuid.uuid4()),
            run_id or str(uuid.uuid4()),
            1,
            RiskPolicy.SAFE,
            scheduled,
            (time.perf_counter() - started) * 1000,
            cursor,
            0,
            0,
        )

File: controlplane/internal/scheduler/scheduler.go
package scheduler

import (
	"fmt"
	"math"
	"math/rand"
	"sort"
	"time"

	"github.com/google/uuid"
	"github.com/newrohansinha/ML-CI-Test-Conflict-Detection-Scheduler/controlplane/internal/types"
)

type Parameters struct {
	RiskWeight       float64
	HardThreshold    float64
	UtilizationBias  float64
	Lookahead        int
	RefinementRounds int
}

func ParametersFor(policy types.RiskPolicy) Parameters {
	switch policy {
	case types.Aggressive:
		return Parameters{0.55, 0.985, 1.5, 8, 2}
	case types.Safe:
		return Parameters{7.5, 0.68, 0.5, 32, 6}
	default:
		return Parameters{2, 0.90, 1, 16, 4}
	}
}

type RiskIndex struct{ predictions map[string]types.Prediction }

func NewRiskIndex(predictions []types.Prediction) (*RiskIndex, error) {
	index := &RiskIndex{predictions: make(map[string]types.Prediction, len(predictions))}
	for _, prediction := range predictions {
		if err := prediction.Normalize(); err != nil {
			return nil, err
		}
		key := types.PairKey(prediction.TestA, prediction.TestB)
		if current, exists := index.predictions[key]; !exists || prediction.Probability > current.Probability {
			index.predictions[key] = prediction
		}
	}
	return index, nil
}

func (r *RiskIndex) Probability(a, b string) float64 {
	if a == b {
		return 1
	}
	return r.predictions[types.PairKey(a, b)].Probability
}

type Scheduler struct {
	Workers    int
	Policy     types.RiskPolicy
	Seed       int64
	Parameters Parameters
}

func New(workers int, policy types.RiskPolicy, seed int64) (*Scheduler, error) {
	if workers <= 0 {
		return nil, fmt.Errorf("worker count must be positive")
	}
	if !policy.Valid() {
		return nil, fmt.Errorf("invalid risk policy %q", policy)
	}
	return &Scheduler{workers, policy, seed, ParametersFor(policy)}, nil
}

type lane struct {
	tests []types.ScheduledTest
	end   float64
}

func (s *Scheduler) Build(runID string, tests []types.Test, predictions []types.Prediction) (types.Schedule, error) {
	started := time.Now()
	risk, err := NewRiskIndex(predictions)
	if err != nil {
		return types.Schedule{}, err
	}
	if runID == "" {
		runID = uuid.NewString()
	}
	remaining := make(map[string]types.Test, len(tests))
	tie := make(map[string]float64, len(tests))
	randomizer := rand.New(rand.NewSource(s.Seed))
	for _, test := range tests {
		if err := test.Validate(); err != nil {
			return types.Schedule{}, fmt.Errorf("invalid test %s: %w", test.ID, err)
		}
		if _, exists := remaining[test.ID]; exists {
			return types.Schedule{}, fmt.Errorf("duplicate test ID %s", test.ID)
		}
		if test.DurationEMA <= 0 {
			test.DurationEMA = 1
		}
		remaining[test.ID], tie[test.ID] = test, randomizer.Float64()
	}
	lanes := make([]lane, s.Workers)
	for len(remaining) > 0 {
		worker := earliestLane(lanes)
		start := lanes[worker].end
		candidates := make([]types.Test, 0, len(remaining))
		for _, test := range remaining {
			candidates = append(candidates, test)
		}
		sort.Slice(candidates, func(i, j int) bool {
			if candidates[i].DurationEMA == candidates[j].DurationEMA {
				return tie[candidates[i].ID] < tie[candidates[j].ID]
			}
			return candidates[i].DurationEMA > candidates[j].DurationEMA
		})
		if len(candidates) > s.Parameters.Lookahead {
			candidates = candidates[:s.Parameters.Lookahead]
		}
		best, bestCost := candidates[0], math.Inf(1)
		for _, candidate := range candidates {
			probability, _ := placementRisk(candidate.ID, start, start+candidate.DurationEMA, lanes, worker, risk)
			cost := start - s.Parameters.UtilizationBias*candidate.DurationEMA + s.Parameters.RiskWeight*probability*candidate.DurationEMA
			if probability >= s.Parameters.HardThreshold {
				cost += 1000
			}
			if cost < bestCost {
				best, bestCost = candidate, cost
			}
		}
		probability, reasons := placementRisk(best.ID, start, start+best.DurationEMA, lanes, worker, risk)
		if probability >= s.Parameters.HardThreshold {
			if next, ok := earliestConflictingEnd(best.ID, start, lanes, worker, risk, s.Parameters.HardThreshold); ok {
				start = next
				probability, reasons = placementRisk(best.ID, start, start+best.DurationEMA, lanes, worker, risk)
			}
		}
		scheduled := types.ScheduledTest{
			TestID: best.ID, NodeID: best.NodeID, Worker: worker,
			EstimatedStart: start, EstimatedEnd: start + best.DurationEMA,
			EstimatedDuration: best.DurationEMA, RiskCost: probability, Reasons: reasons,
		}
		lanes[worker].tests = append(lanes[worker].tests, scheduled)
		lanes[worker].end = scheduled.EstimatedEnd
		delete(remaining, best.ID)
	}
	if len(tests) <= 64 {
		for round := 0; round < s.Parameters.RefinementRounds; round++ {
			if !refine(lanes, risk, s.Parameters.RiskWeight) {
				break
			}
		}
	}
	flattened := make([]types.ScheduledTest, 0, len(tests))
	makespan := 0.0
	for _, lane := range lanes {
		flattened = append(flattened, lane.tests...)
		makespan = math.Max(makespan, lane.end)
	}
	sort.Slice(flattened, func(i, j int) bool {
		if flattened[i].EstimatedStart == flattened[j].EstimatedStart {
			return flattened[i].Worker < flattened[j].Worker
		}
		return flattened[i].EstimatedStart < flattened[j].EstimatedStart
	})
	result := types.Schedule{
		ID: uuid.NewString(), RunID: runID, Workers: s.Workers, Policy: s.Policy,
		Tests: flattened, SchedulerLatencyMS: float64(time.Since(started).Microseconds()) / 1000,
		ExpectedMakespan: makespan, ExpectedRisk: scheduleRisk(flattened, risk), Seed: s.Seed,
	}
	return result, result.Validate()
}

func earliestLane(lanes []lane) int {
	best := 0
	for i := 1; i < len(lanes); i++ {
		if lanes[i].end < lanes[best].end {
			best = i
		}
	}
	return best
}

func placementRisk(testID string, start, end float64, lanes []lane, worker int, risk *RiskIndex) (float64, []string) {
	safe := 1.0
	reasons := []string{}
	for index, current := range lanes {
		if index == worker {
			continue
		}
		for _, other := range current.tests {
			overlap := math.Max(0, math.Min(end, other.EstimatedEnd)-math.Max(start, other.EstimatedStart))
			if overlap == 0 {
				continue
			}
			probability := risk.Probability(testID, other.TestID)
			exposure := math.Min(1, overlap/math.Max(.001, math.Min(end-start, other.EstimatedDuration)))
			safe *= 1 - probability*exposure
			if probability >= .25 {
				reasons = append(reasons, fmt.Sprintf("%s: %.1f%% conflict risk", other.NodeID, probability*100))
			}
		}
	}
	return 1 - safe, reasons
}

func earliestConflictingEnd(testID string, now float64, lanes []lane, worker int, risk *RiskIndex, threshold float64) (float64, bool) {
	best := math.Inf(1)
	found := false
	for index, current := range lanes {
		if index == worker {
			continue
		}
		for _, other := range current.tests {
			if other.EstimatedStart <= now && now < other.EstimatedEnd && risk.Probability(testID, other.TestID) >= threshold {
				best = math.Min(best, other.EstimatedEnd)
				found = true
			}
		}
	}
	return best, found
}

func scheduleRisk(tests []types.ScheduledTest, risk *RiskIndex) float64 {
	total := 0.0
	for i, left := range tests {
		for _, right := range tests[i+1:] {
			if left.Worker == right.Worker {
				continue
			}
			overlap := math.Max(0, math.Min(left.EstimatedEnd, right.EstimatedEnd)-math.Max(left.EstimatedStart, right.EstimatedStart))
			total += overlap * risk.Probability(left.TestID, right.TestID)
		}
	}
	return total
}

func objective(lanes []lane, risk *RiskIndex, weight float64) float64 {
	all := []types.ScheduledTest{}
	makespan := 0.0
	for _, lane := range lanes {
		all = append(all, lane.tests...)
		makespan = math.Max(makespan, lane.end)
	}
	return makespan + weight*scheduleRisk(all, risk)
}
func reflow(target *lane, worker int) {
	cursor := 0.0
	for index := range target.tests {
		target.tests[index].Worker = worker
		target.tests[index].EstimatedStart = cursor
		cursor += target.tests[index].EstimatedDuration
		target.tests[index].EstimatedEnd = cursor
	}
	target.end = cursor
}
func refine(lanes []lane, risk *RiskIndex, weight float64) bool {
	changed := false
	baseline := objective(lanes, risk, weight)
	for worker := range lanes {
		for i := 0; i+1 < len(lanes[worker].tests); i++ {
			lanes[worker].tests[i], lanes[worker].tests[i+1] = lanes[worker].tests[i+1], lanes[worker].tests[i]
			reflow(&lanes[worker], worker)
			candidate := objective(lanes, risk, weight)
			if candidate+1e-9 < baseline {
				baseline = candidate
				changed = true
			} else {
				lanes[worker].tests[i], lanes[worker].tests[i+1] = lanes[worker].tests[i+1], lanes[worker].tests[i]
				reflow(&lanes[worker], worker)
			}
		}
	}
	return changed
}
